src/eh1/io.rs

```diff
@@ -0,0 +1,617 @@
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(()),
+        }
+    }
+}
+
+impl Read for Mock {
+    fn read(&mut self, buffer: &mut [u8]) -> Result<usize, Self::Error> {
+        let w = self.next().expect("no expectation for io::read call");
+        assert_eq!(w.expected_mode, Mode::Read, "io::read unexpected mode");
+        buffer.copy_from_slice(&w.response);
+
+        match w.expected_err {
+            Some(err) => Err(err),
+            None => Ok(buffer.len()),
+        }
+    }
+}
+
+impl Seek for Mock {
+    fn seek(&mut self, pos: SeekFrom) -> Result<u64, Self::Error> {
+        let w = self.next().expect("no expectation for io::seek call");
+
+        if let Mode::Seek(expected_pos) = w.expected_mode {
+            assert_eq!(expected_pos, pos, "io::seek unexpected mode");
+
+            let ret_offset: u64 = u64::from_be_bytes(w.response.try_into().unwrap());
+            match w.expected_err {
+                Some(err) => Err(err),
+                None => Ok(ret_offset),
+            }
+        } else {
+            panic!("unexpected seek mode");
+        }
+    }
+}
+
+impl WriteReady for Mock {
+    fn write_ready(&mut self) -> Result<bool, Self::Error> {
+        let w = self
+            .next()
+            .expect("no expectation for io::write_ready call");
+
+        match w.expected_mode {
+            Mode::WriteReady(ready) if w.expected_err.is_none() => Ok(ready),
+            Mode::WriteReady(_) if w.expected_err.is_some() => Err(w.expected_err.unwrap()),
+            _ => panic!("unexpected write_ready mode"),
+        }
+    }
+}
+
+impl ReadReady for Mock {
+    fn read_ready(&mut self) -> Result<bool, Self::Error> {
+        let w = self.next().expect("no expectation for io::read_ready call");
+
+        match w.expected_mode {
+            Mode::ReadReady(ready) if w.expected_err.is_none() => Ok(ready),
+            Mode::ReadReady(_) if w.expected_err.is_some() => Err(w.expected_err.unwrap()),
+            _ => panic!("unexpected read_ready mode"),
+        }
+    }
+}
+
+impl BufRead for Mock {
+    fn fill_buf(&mut self) -> Result<&[u8], Self::Error> {
+        let w = self.next().expect("no expectation for io::fill_buf call");
+        assert_eq!(
+            w.expected_mode,
+            Mode::FillBuff,
+            "io::fill_buf unexpected mode"
+        );
+
+        let response_vec = w.response;
+
+        match w.expected_err {
+            Some(err) => Err(err),
+
+            // SAFETY : This is a memory leak. This is done on purpose to allow for a return
+            // of a slice (pointer) to the internal buffer, which doesn't exist in mock implementation.
+            // This way, after each call, response is going to be put on the heap and ref is going to be returned, without freeing this memory.
+            // For mocking purposes this should be an acceptable solution.
+            None => Ok(Box::leak(response_vec.into_boxed_slice())),
```

Comment:
I think this solution is fine for a mocking library.

However, another option would be to add a buffer to the mock. Since most mocks use an alias to `Generic<T>`, this isn't trivial to add. But one way might be to parametrize the `Generic` type and add a general-purpose data field, something like this:

```rust
#[derive(Debug, Clone)]
pub struct Generic<T: Clone + Debug + PartialEq, B = ()> {
    expected: Arc<Mutex<VecDeque<T>>>,
    done_called: Arc<Mutex<DoneCallDetector>>,
    data: B,
}
```

This way, a mock could add associated data to the generic mock. In this case, it would be a generic buffer of a certain size. But it could also be a struct, depending on the mock type.